clang/lib/Sema/SemaExpr.cpp

```diff
@@ -7813,6 +7813,32 @@ Sema::BuildCompoundLiteralExpr(SourceLocation LParenLoc, TypeSourceInfo *TInfo,
                                            diag::err_variable_object_no_init))
         return ExprError();
     }
+  } else if (LangOpts.C23 &&
```

Comment:
I don't think these changes are quite sufficient because I think we need to handle more cases than just compound literals. Consider:
```
constexpr typeof(struct s *) x = 0; // declares `s` which is not an ordinary identifier
constexpr struct S { int a, b; } y = { 0 }; // declares `S`, `a`, and `b`, none of which are ordinary identifiers
constexpr int a = 0, b = 0; // declares `a` and `b` as ordinary identifiers
auto c = (struct T { int x, y; }){0, 0}; // declares `T`, `x`, and `y`, none of which are ordinary identifiers
constexpr int (*fp)(struct X { int x; } val) = 0; // declares `X` and `x` which are not ordinary identifiers
```
(https://godbolt.org/z/Pq87aMnch)

Reply:
3 out of these 5 cases are not conforming curently:
```cpp
constexpr typeof(struct s *) x = 0;               // Wrong, no underspecifed diagnostics
constexpr struct S { int a, b; } y = { 0 };       // Wrong, no underspecifed diagnostics
constexpr int a = 0, b = 0;                       // Ok, no underspecifed diagnostics
auto c = (struct T { int x, y; }){0, 0};          // Ok, `struct T` diagnosed as underspecifed
constexpr int (*fp)(struct X { int x; } val) = 0; // Wrong, no underspecifed diagnostics but dignoses warning: `declaration of 'struct X' will not be visible outside of this function`
```
I need to do more work outside of compound literals.